### DISCUSSION TITLE
How to extend custom Adapter types?

### DISCUSSION THREAD

**Discussion by ugglr** [nextauthjs/next-auth#6245] [Help]
I've implemented my own custom adapter for a stand-alone backend, but I just cannot get the types to work correctly. 

How to extend the types of `Adapter`? (I only support Github login)

here's some code for my use case:

```
export default function BackendAdapter({ client }: Props): Adapter {
  return {
    async createUser(user) {
      console.log("Creating a new user 🚀");
      try {
        const { ghId, email } = user;

        // Makes the call to my stand-alone BE
        const {
          data: { createUser },
        } = await client.mutate({
          mutation: M_CREATE_USER,
          variables: { ghId, email },
        });

        return {
          id: createUser.id,
          ghId,
          isAdmin: createUser.isAdmin,
          email: createUser.email,
          stars: createUser.stars,
          sessionToken: null,
        };
      } catch (err) {
        return;
      }
    },
```

I've tried to extend the types in `next-auth.d.ts` as the docs suggest but it does not seem to work for the `Adapter` functions??

for instance:

```
import NextAuth from "next-auth";
import { AdapterUser } from "next-auth/adapters";

declare module "next-auth" {
  interface User {
    id: string;
    ghId: number;
    isAdmin: boolean;
    email: string;
    sessionToken: string | null;
  }
  interface Session {
    user: User;
  }

  interface Adapter {
    createUser(User: AdapterUser): () => Promise<User> | undefined;
  }
}
```

It's been quite difficult so far to get to this point, the docs do not talk much about having a separate backend running, it mostly focuses on running straight into a database. 

Any help would be appreciated 🚀

**Comment by NazCodeland** [comment]
Hey did you ever figure out how to extend the AdapterUser interface? @ugglr

**Comment by bsplosion** [comment]
I did something like this to extend the types - this allowed the `AdapterUser` object to pick up the attributes correctly and also allows type hinting of parameters in case you leverage custom functions to handle these values.

The most important realization is that `AdapterUser` extends `User` so you can set your custom user attributes once in here, tell `Session` to use them, and everything else should just "click" in place, no further interfaces or types required.

```
import NextAuth, { DefaultSession, DefaultUser } from 'next-auth';

declare module 'next-auth' {
  /**
   * Leveraged by session callback's user object (AdapterUser extends User)
   */
  export interface User extends DefaultUser {
    /** Define any user-specific variables here to make them available to other code inferences */
    ghId: number;
    isAdmin: boolean;
    // Any other attributes you need from either your User table columns or additional fields during a session callback
  }

  /**
   * Returned by `useSession`, `getSession` and received as a prop on the `SessionProvider` React Context
   */
  export interface Session {
    user: User;
  }
}
```

**Reply by MathiasSchjoedt-Bavngaard** [reply]
Thank you! This helped me in my search for "change AdapterUser NextAuth"

**Comment by lingo-anshul** [comment]
You can put this in `types/adapters.d.ts` file -

```
import type { AdapterUser as BaseAdapterUser } from "next-auth/adapters";

declare module "@auth/core/adapters" {
  interface AdapterUser extends BaseAdapterUser {
    ghId: number;
    isAdmin: boolean;
  }
}
```
